gabxgonz/cat-game-collab
# Pickups

## Mouse Toy & Pickup

This primarily adds a `Pickup` class that inherits from `Interactable`.
The goal was to add a mouse toy that could be picked up by the player.
![2021-09-04-pickup-mouse](https://user-images.githubusercontent.com/75653211/139596369-81627b1b-f369-4ed5-a9bf-6d4942676ea6.gif)

### To Test

* Add the `Mouse Toy` prefab to the scene.
* Walk near the mouse toy and press the Mouth key (R at the time of writing).
* The cat should pick up the `Mouse Toy` and play the Eating animation.

### Implementation
Here's a 'sequence' diagram of the main change
![Screen Shot 2021-10-30 at 4 27 30 PM](https://user-images.githubusercontent.com/75653211/139561128-20d5f9d9-e6c6-48c0-aaa6-49910dad4ec8.png)

When the `PlayerController` calls `HandleInteraction`, it now passes a `Transform` that represents where a pickup should attach to the mouth. If we end up needing more attachment points for Interactions, we can either add more arguments or pass a `Dictionary` instead.

The `Interactable` class now has an `overloaded` method for `MouthInteraction`, meaning there is more than one function of the same name but with a different argument signature. The two signatures are:

* `HandleInteraction()`
* `HandleInteraction(Transform? attachmentPoint)`

C# will select the right function to use based on what arguments you pass along.

Each instance of the `Pickup` class keeps track of whether or not it's currently being held. If it isn't, it parents itself to the given attachment point and disables gravity/enables Kinematics on it's `Rigidbody` since it's now being controlled programmatically. If it is being held, it does the inverse of this to allow it to fall naturally to the ground.

Finally, it overrides the animation selected by the `Interactable` class. This is so that we may play a different `PlayerAnimation` (None) when the item is dropped than when it's picked up (Eat).
